app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java

```diff
@@ -1,40 +1,54 @@
+import java.util.Map;
+import java.util.Map.Entry;
+
 import io.appium.uiautomator2.handler.request.SafeRequestHandler;
 import io.appium.uiautomator2.http.AppiumResponse;
 import io.appium.uiautomator2.http.IHttpRequest;
-import io.appium.uiautomator2.model.AllowInvisibleElements;
-import io.appium.uiautomator2.model.CompressedLayoutHierarchy;
 import io.appium.uiautomator2.model.Session;
-import io.appium.uiautomator2.model.WaitForIdleTimeout;
+import io.appium.uiautomator2.model.settings.AllowInvisibleElements;
+import io.appium.uiautomator2.model.settings.EnableNotificationListener;
+import io.appium.uiautomator2.model.settings.CompressedLayoutHierarchy;
+import io.appium.uiautomator2.model.settings.ISetting;
+import io.appium.uiautomator2.model.settings.WaitForIdleTimeout;
 import io.appium.uiautomator2.server.WDStatus;
 import io.appium.uiautomator2.utils.Logger;
-import java.util.Map;
 
 public class UpdateSettings extends SafeRequestHandler {
 
+    private static final Map<String, Class<? extends ISetting>> SETTINGS = new HashMap<String, Class<? extends ISetting>>() {
+        {
+            put(AllowInvisibleElements.SETTING_NAME, AllowInvisibleElements.class);
+            put(CompressedLayoutHierarchy.SETTING_NAME, CompressedLayoutHierarchy.class);
+            put(WaitForIdleTimeout.SETTING_NAME, WaitForIdleTimeout.class);
+            put(EnableNotificationListener.SETTING_NAME, EnableNotificationListener.class);
+        }
+    };
+
     public UpdateSettings(String mappedUri) {
         super(mappedUri);
     }
 
     @Override
     public AppiumResponse safeHandle(IHttpRequest request) {
-
         try {
             Map<String, Object> settings = getPayload(request, "settings");
-            Session.capabilities.putAll(settings);
             Logger.debug("Update settings: " + settings.toString());
-
-            if (settings.containsKey(AllowInvisibleElements.SETTING_NAME)) {
-                AllowInvisibleElements.updateSetting((boolean) settings.get(AllowInvisibleElements.SETTING_NAME));
-            }
-
-            if (settings.containsKey(CompressedLayoutHierarchy.SETTING_NAME)) {
-                CompressedLayoutHierarchy.updateSetting((boolean) settings.get(CompressedLayoutHierarchy.SETTING_NAME));
-            }
-
-            if (settings.containsKey(WaitForIdleTimeout.SETTING_NAME)) {
-                WaitForIdleTimeout.updateSetting((int) settings.get(WaitForIdleTimeout.SETTING_NAME));
+            for (Entry<String, Object> entry : settings.entrySet()) {
+                String settingName = entry.getKey();
+                Object settingValue = entry.getValue();
+                ISetting setting = getSetting(settingName);
+                if (setting != null) {
+                    setting.updateSetting(settingValue);
+                    Session.capabilities.put(settingName, settingValue);
+                } else {
+                    Logger.error(String.format("Setting '%s' is not supported.", entry.getKey()));
```

Comment:
shall we throw an exception instead?

Reply:
From my point of view we can just skip unsupported setting like unsupported desired capabilities.

Reply:
unsupported caps are visible in the server log (and even there not so many people notice this). How do you think - how many of them are going to check device logs?

Reply:
ok

Reply:
This all goes through the driver, so validation/logging can happen there. Or the calling code could catch an error and log it, while continuing to function normally.